pkg/cmd/cli/cmd/deploy.go

```diff
@@ -0,0 +1,181 @@
+	cmd.Flags().BoolVar(&retryDeploy, "retry", false, "Retry the latest failed deployment.")
+
+	return cmd
+}
+
+// deployCommandClient abstracts access to the API server.
+type deployCommandClient interface {
+	GetDeployment(namespace, name string) (*kapi.ReplicationController, error)
+	UpdateDeploymentConfig(*deployapi.DeploymentConfig) (*deployapi.DeploymentConfig, error)
+	UpdateDeployment(*kapi.ReplicationController) (*kapi.ReplicationController, error)
+}
+
+// deployLatestCommand can launch new deployments.
+type deployLatestCommand struct {
+	client deployCommandClient
+}
+
+// deploy launches a new deployment unless there's already a deployment
+// process in progress for config.
+func (c *deployLatestCommand) deploy(config *deployapi.DeploymentConfig, out io.Writer) error {
+	deploymentName := deployutil.LatestDeploymentNameForConfig(config)
+	deployment, err := c.client.GetDeployment(config.Namespace, deploymentName)
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return err
+		}
+	} else {
+		// Reject attempts to start a concurrent deployment.
+		status := statusFor(deployment)
+		if status != deployapi.DeploymentStatusComplete && status != deployapi.DeploymentStatusFailed {
+			return fmt.Errorf("#%d is already in progress (%s)", config.LatestVersion, status)
+		}
+	}
+
+	config.LatestVersion++
+	_, err = c.client.UpdateDeploymentConfig(config)
+	if err == nil {
+		fmt.Fprintf(out, "deployed #%d\n", config.LatestVersion)
+	}
+	return err
+}
+
+// retryDeploymentCommand can retry failed deployments.
+type retryDeploymentCommand struct {
+	client deployCommandClient
+}
+
+// retry resets the status of the latest deployment to New, which will cause
+// the deployment to be retried. An error is returned if the deployment is not
+// currently in a failed state.
+func (c *retryDeploymentCommand) retry(config *deployapi.DeploymentConfig, out io.Writer) error {
+	if config.LatestVersion == 0 {
+		return fmt.Errorf("no failed deployments found for %s/%s", config.Namespace, config.Name)
+	}
+	deploymentName := deployutil.LatestDeploymentNameForConfig(config)
+	deployment, err := c.client.GetDeployment(config.Namespace, deploymentName)
+	if err != nil {
+		return err
+	}
+
+	status := statusFor(deployment)
+
+	if status != deployapi.DeploymentStatusFailed {
+		return fmt.Errorf("#%d is %s; only failed deployments can be retried", config.LatestVersion, status)
+	}
+
+	deployment.Annotations[deployapi.DeploymentStatusAnnotation] = string(deployapi.DeploymentStatusNew)
+	_, err = c.client.UpdateDeployment(deployment)
+	if err == nil {
+		fmt.Fprintf(out, "retried #%d\n", config.LatestVersion)
+	}
+	return err
+}
+
+func statusFor(deployment *kapi.ReplicationController) deployapi.DeploymentStatus {
+	return deployapi.DeploymentStatus(deployment.Annotations[deployapi.DeploymentStatusAnnotation])
+}
+
+// deployCommandClientImpl is a pluggable deployCommandClient.
+type deployCommandClientImpl struct {
```

Comment:
I can revisit the interface usage the next time we come in here to add more stuff.